### PR TITLE
bpo-40025:The behavior is the same when `_generate_next_value_` is defined afte…

### PR DESCRIPTION
**Pull request by hongweipeng**
…r or before members

https://bugs.python.org/issue40025

### CODE REVIEW

File: Lib/enum.py [hunk 1]
```diff
@@ -104,10 +101,12 @@ def __setitem__(self, key, value):
                 # enum overwriting a descriptor?
                 raise TypeError('%r already defined as: %r' % (key, self[key]))
             if isinstance(value, auto):
-                self._auto_called = True
-                if value.value == _auto_null:
-                    value.value = self._generate_next_value(key, 1, len(self._member_names), self._last_values[:])
-                value = value.value
+                _value = value.value
+                if _value == _auto_null and self._auto_called:
```
Comment (ethanfurman) [inline Lib/enum.py]: `self._auto_called` is never set to true inside `_EnumDict`, so this test will never succeed.

File: Lib/enum.py [hunk 2]
```diff
@@ -148,6 +147,14 @@ def __new__(metacls, cls, bases, classdict):
...
         # save enum items into separate mapping so they don't get baked into
         # the new class
+        for i,name in enumerate(classdict._member_names):
+            value = classdict[name]
+            if isinstance(value, auto):
+                if value.value == _auto_null:
+                    value.value = classdict._generate_next_value(name, 1, len(classdict._member_names[:i]), classdict._last_values[:i])
+                super(type(classdict), classdict).__setitem__(name, value.value)
+                classdict._last_values[i] = value.value
+        super(type(classdict), classdict).__setitem__('_auto_called', True)
```
Comment (ethanfurman) [inline Lib/enum.py]: Setting `_auto_called` here has no effect.

File: Lib/enum.py [hunk 3]
```diff
@@ -148,6 +147,14 @@ def __new__(metacls, cls, bases, classdict):
 
         # save enum items into separate mapping so they don't get baked into
         # the new class
+        for i,name in enumerate(classdict._member_names):
+            value = classdict[name]
+            if isinstance(value, auto):
+                if value.value == _auto_null:
```
Comment (ethanfurman) [inline Lib/enum.py]: Will `value` ever be an `auto` and *not* be `_auto_null`?

**Review by ethanfurman — CHANGES_REQUESTED** [review]
I like the idea, but it needs more work.  Currently, even if `_generate_next_value_` is defined before any members it still will not be called until after the entire `Enum` is defined and given to `EnumMeta.__new__` for final processing, which means that somebody wanting to do interesting things with previous member values would not be able to.

In `_EnumDict` you'll need to track if `auto()` has been seen yet, if `_generate_next_value_` was defined before any uses of `auto()` were, and call `_generate_next_value_` within `_EnumDict` if and only if `_generate_next_value_` was defined before *all* uses of `auto()`.

So this should work:

    class Test1(Enum):
        def _generate_next_value(...):
            return 12
        first = auto()
        second = first + 9

and this should fail:

    class Test2(Enum):
        first = auto()
        def _generate_next_value(...):
            return 12
        second = first + 9

Also, make sure you have spaces after commas, etc.

File: Lib/enum.py (hunk 2 above)
Comment (hongweipeng) [reply Lib/enum.py]:
```
class Dupes(Enum):
    first = primero = auto()
```
In this case,`primero` is not `_auto_null`.

Comment (ethanfurman) [reply Lib/enum.py]: Ah, you're right.  Thanks.

**Comment by hongweipeng** [comment]
```
class Test2(Enum):
    first = auto()
    def _generate_next_value_(...):
        return 12
    second = first + 9
```
I think this should not fail, as long as `second` is defined after `_generate_next_value_`. I added this example to the unit test.

I have made the requested changes; please review again.

**Comment by ethanfurman** [comment]
I had hoped to incorporate this idea when I had time to look at it further; unfortunately, now that I have I do not see a way to support the following use-case (which I believe would be much more common):

    class I(Enum):
        first = auto()
        second = first + 2

HongWeipeng, thank you for trying.

**Pull request closed without merging**